File: packages/sdk/python/roboviz/multi_robot.py
"""
RoboViz Multi-Robot Coordination

Multi-robot group management: synchronized trajectories, shared workspace,
coordination modes.

Example:
    >>> import roboviz as rv
    >>> rv.init()
    >>> group = rv.RobotGroup(rv.get_instance(), "cell-1", ["robot-a", "robot-b"])
    >>> group.set_coordination_type("synchronized")
    >>> group.play_sync_trajectory(trajectories)
"""

from typing import List, Dict, Any, Optional, TYPE_CHECKING

from .types import TrajectoryData

if TYPE_CHECKING:
    from .core import RoboViz


class RobotGroup:
    """
    Multi-robot coordination group.

    Args:
        viz: Active RoboViz instance.
        group_id: Unique group identifier.
        robot_ids: List of robot IDs in the group.
    """

    def __init__(self, viz: "RoboViz", group_id: str, robot_ids: Optional[List[str]] = None):
        self._viz = viz
        self._group_id = group_id
        self._robot_ids = list(robot_ids or [])

        # Create group on viewer side
        self._viz._send_request("multiRobot.createGroup", {
            "groupId": group_id,
            "robotIds": self._robot_ids,
        })

    @property
    def group_id(self) -> str:
        return self._group_id

    @property
    def robot_ids(self) -> List[str]:
        return list(self._robot_ids)

    # =========================================================================
    # Group Management
    # =========================================================================

    def add_robot(self, robot_id: str) -> None:
        """Add a robot to the group."""
        self._viz._send_notification("multiRobot.addToGroup", {
            "groupId": self._group_id,
            "robotId": robot_id,
        })
        self._robot_ids.append(robot_id)

    def remove_robot(self, robot_id: str) -> None:
        """Remove a robot from the group."""
        self._viz._send_notification("multiRobot.removeFromGroup", {
            "groupId": self._group_id,
            "robotId": robot_id,
        })
        if robot_id in self._robot_ids:
            self._robot_ids.remove(robot_id)

    def set_coordination_type(self, coord_type: str) -> None:
        """
        Set coordination mode.

        Args:
            coord_type: 'independent', 'synchronized', 'leader_follower',
                       'cooperative'.
        """
        self._viz._send_notification("multiRobot.setCoordinationType", {
            "groupId": self._group_id,
            "type": coord_type,
        })

    def set_master(self, robot_id: str) -> None:
        """Set the leader robot for leader-follower mode."""
        self._viz._send_notification("multiRobot.setMaster", {
            "groupId": self._group_id,
            "robotId": robot_id,
        })

    # =========================================================================
    # Synchronized Playback
    # =========================================================================

    def play_sync_trajectory(
        self,
        trajectories: Dict[str, TrajectoryData],
        speed: float = 1.0,
    ) -> None:
        """
        Play synchronized trajectories for all robots.

        Args:
            trajectories: Mapping of robot_id → TrajectoryData.
            speed: Playback speed multiplier.
        """
        traj_dict = {rid: t.to_dict() for rid, t in trajectories.items()}
        self._viz._send_notification("multiRobot.playSyncTrajectory", {
            "groupId": self._group_id,
            "trajectories": traj_dict,
            "speed": speed,
        })

    def pause(self) -> None:
        """Pause synchronized playback."""
        self._viz._send_notification("multiRobot.pause", {"groupId": self._group_id})

    def stop(self) -> None:
        """Stop synchronized playback."""
        self._viz._send_notification("multiRobot.stop", {"groupId": self._group_id})

    def resume(self) -> None:
        """Resume synchronized playback."""
        self._viz._send_notification("multiRobot.resume", {"groupId": self._group_id})

    # =========================================================================
    # Joint Control
    # =========================================================================

    def set_all_joints(self, joints_map: Dict[str, List[float]]) -> None:
        """
        Set joint angles for all robots simultaneously.

        Args:
            joints_map: Mapping of robot_id → joint angles.
        """
        self._viz._send_notification("multiRobot.setAllJoints", {
            "groupId": self._group_id,
            "joints": joints_map,
        })

    def get_all_joints(self) -> Dict[str, List[float]]:
        """
        Get joint angles for all robots.

        Returns:
            Mapping of robot_id → current joint angles.
        """
        result = self._viz._send_request("multiRobot.getAllJoints", {
            "groupId": self._group_id,
        })
        return result.get("joints", {})

    # =========================================================================
    # Workspace Analysis
    # =========================================================================

    def calculate_shared_workspace(
        self,
        resolution: int = 20,
    ) -> Dict[str, Any]:
        """
        Calculate the shared workspace between all robots in the group.

        Args:
            resolution: Sampling resolution per axis.

        Returns:
            Dict with 'overlapVolume', 'overlapPoints', 'robotWorkspaces'.
        """
        return self._viz._send_request("multiRobot.calculateSharedWorkspace", {
            "groupId": self._group_id,
            "resolution": resolution,
        })

    def check_inter_robot_collision(self) -> Dict[str, Any]:
        """
        Check for collisions between robots in the group.

        Returns:
            Dict with 'hasCollision', 'pairs'.
        """
        return self._viz._send_request("multiRobot.checkInterRobotCollision", {
            "groupId": self._group_id,
        })

    # =========================================================================
    # Lifecycle
    # =========================================================================

    def dispose(self) -> None:
        """Remove the group from the viewer."""
        self._viz._send_notification("multiRobot.removeGroup", {"groupId": self._group_id})

    def __repr__(self) -> str:
        return f"RobotGroup(id={self._group_id!r}, robots={self._robot_ids})"

File: packages/sdk/rust/src/multi_robot.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use serde_json::{json, Value};

use crate::client::{send_request, send_notify, ClientInner};
use crate::error::Result;

/// A group of robots that can be controlled together.
///
/// Supports synchronised trajectory playback and bulk joint commands.
pub struct RobotGroup {
    inner: Arc<ClientInner>,
    timeout: Duration,
    group_id: String,
}

impl RobotGroup {
    /// Create a new robot group with the given member robots.
    ///
    /// This registers the group on the server.
    pub async fn new(
        client: &crate::client::RoboVizClient,
        group_id: impl Into<String>,
        robot_ids: &[&str],
    ) -> Result<Self> {
        let inner = Arc::clone(client.inner());
        let timeout = client.default_timeout();
        let group_id = group_id.into();

        send_request(
            &inner,
            timeout,
            "multiRobot.createGroup",
            json!({
                "groupId": group_id,
                "robotIds": robot_ids,
            }),
        )
        .await?;

        Ok(Self { inner, timeout, group_id })
    }

    /// Add a robot to this group.
    pub async fn add_robot(&self, robot_id: &str) -> Result<()> {
        send_request(
            &self.inner,
            self.timeout,
            "multiRobot.addRobot",
            json!({
                "groupId": self.group_id,
                "robotId": robot_id,
            }),
        )
        .await?;
        Ok(())
    }

    /// Remove a robot from this group.
    pub async fn remove_robot(&self, robot_id: &str) -> Result<()> {
        send_request(
            &self.inner,
            self.timeout,
            "multiRobot.removeRobot",
            json!({
                "groupId": self.group_id,
                "robotId": robot_id,
            }),
        )
        .await?;
        Ok(())
    }

    /// Set joint angles for every robot in the group at once.
    ///
    /// `joints_map` maps robot IDs to their joint-angle vectors.
    pub async fn set_all_joints(
        &self,
        joints_map: &HashMap<String, Vec<f64>>,
    ) -> Result<()> {
        send_notify(
            &self.inner,
            "multiRobot.setAllJoints",
            json!({
                "groupId": self.group_id,
                "jointsMap": joints_map,
            }),
        )
        .await
    }

    /// Get the current joint angles of every robot in the group.
    pub async fn get_all_joints(&self) -> Result<HashMap<String, Vec<f64>>> {
        let result = send_request(
            &self.inner,
            self.timeout,
            "multiRobot.getAllJoints",
            json!({ "groupId": self.group_id }),
        )
        .await?;
        Ok(serde_json::from_value(result)?)
    }

    /// Play synchronised trajectories across all robots.
    ///
    /// `trajectories` is a JSON value mapping robot IDs to trajectory data.
    pub async fn play_sync_trajectory(
        &self,
        trajectories: Value,
        speed: f64,
    ) -> Result<()> {
        send_notify(
            &self.inner,
            "multiRobot.playSyncTrajectory",
            json!({
                "groupId": self.group_id,
                "trajectories": trajectories,
                "speed": speed,
            }),
        )
        .await
    }

    /// Dispose of the group, releasing server-side resources.
    pub async fn dispose(&self) -> Result<()> {
        send_request(
            &self.inner,
            self.timeout,
            "multiRobot.disposeGroup",
            json!({ "groupId": self.group_id }),
        )
        .await?;
        Ok(())
    }
}
